### ISSUE TITLE
GRPC_VERBOSITY and GRPC_TRACE env vars don't seem to be taken into account when running as a non-root user

### ISSUE THREAD

**Issue by mably** [grpc/grpc#12093]
### What version of gRPC and what language are you using?

It's a NodeJS (7.1.0)  / gRPC (1.4.1) web application connecting to a lnd (lightning network daemon) gRPC server securely using a certifiate.
 
 
### What operating system (Linux, Windows, …) and version?

Ubuntu 16.04 

### What runtime / compiler are you using (e.g. python version or version of gcc)

gcc installed version is 5.4.0
 
### What did you do?

I just defined those two environment variables (GRPC_VERBOSITY and GRPC_TRACE). 
Logs show fine when running the nodejs app as root, but none are showing in the console when running with a non-root user.
And the certificate used to connect to the secured gRPC server doesn't seem to be recognized (works perfectly when run as root user), that's the failing part I'm trying to debug here.
 
### What did you expect to see?
 
Debugging logs.
 
### What did you see instead?
 
Nothing

Thanks for your help.

**Comment by murgatroid99** [comment]
Can you show the exact command lines in question here? Are you sure you're not piping the output somewhere?

**Comment by mably** [comment]
It's a simple `/usr/bin/nodejs <js file> <cl parameters>` with no piping.

**Comment by murgatroid99** [comment]
Can you include the part where you set the environment variables?

**Comment by mably** [comment]
```
#!/bin/bash
# Starts the lncli web app

export DEBUG=lncliweb*

export GRPC_TRACE=transport_security,connectivity_state
export GRPC_VERBOSITY=DEBUG

/usr/bin/nodejs /home/lncliweb/lncli-web/server --serverhost <...>
```

**Comment by murgatroid99** [comment]
In my own tests, I can't reproduce the behavior you're describing.

One possibility I can think of is that you are trying to bind a port that the user `lncliweb` does not have permission to bind. Other than the issue with the logs, does the server seem to work properly?

**Comment by mably** [comment]
Yep, my server as always been running as the `lncliweb` user.

I notice the problems with the logs while trying to debug this PR where I try to add secure credentials to my application:

https://github.com/mably/lncli-web/pull/90

It seems the `process.env["GRPC_SSL_CIPHER_SUITES"] = "HIGH+ECDSA";` code is not taken into account when running as `lncliweb` user, but it works perfectly fine when run as `root`.

**Comment by nicolasnoble** [comment]
This is an interesting issue, really. We do have sort of a weird setup to read environment variables in the C core due to security concerns over their usage, but I don't see how this would fail for you, given how typical your platform is (Ubuntu 16.04).

https://github.com/grpc/grpc/blob/master/src/core/lib/support/env_linux.c#L49

Maybe there's something else about what you're doing that's triggering secure_getenv's fuses ?

https://linux.die.net/man/3/secure_getenv

> Secure execution is required if one of the following conditions was true when the program run by the calling process was loaded:
> 
> * the process's effective user ID did not match its real user ID or the process's effective group ID did not match its real group ID (typically this is the result of executing a set-user-ID or set-group-ID program);
> * the effective capability bit was set on the executable file; or
> * the process has a nonempty permitted capability set.
> 
> Secure execution may also required if triggered by some Linux security modules.
>

**Comment by mably** [comment]
In fact I'm running this command to be able to run my nodejs app on port 80:

```setcap 'cap_net_bind_service=+ep' /usr/bin/nodejs```

Could be related.

What am I supposed to do then?

**Comment by nicolasnoble** [comment]
That's definitely the issue. We really do not want to stop using secure_getenv to avoid cipher suites and ssl root certificates to be tampered with. But this indeed has the broader side effect to nullify some legitimate usages such as privileged ports bindings.

My personal advice however to use a privileged port in a non-root process is to use an iptable rule, such as:

```
iptables -A PREROUTING -t nat -i $EXTERNAL_INTERFACE -p tcp --dport 80 -j REDIRECT --to-port 8080
```

Getting the iptables rule to work properly may sometimes be a bit tricky when it comes to machines with multiple interfaces, especially if it's also serving as a nat for the rest of the internal network, but this rule above would work properly in the case of a simple leaf server.

**Comment by mably** [comment]
I can confirm that removing the special capability everything works fine.

I'll try your iptables solution instead.

Thank you everybody for your help.

**Comment by mattolson** [comment]
I'm running into a similar issue. When I run my node server on my Mac host, GRPC_TRACE works as expected. When I run in a docker container (Debian Jessie) using an unprivileged user, it does not work. When I run in the container using root, it works.

**Comment by mattolson** [comment]
Nevermind, my nodejs executable also had `cap_net_bind_service+ep` set. Note that is is a stock install, or at least I don't know of anything I did to set this capability. Fortunately, in our environment, we use Docker with port mapping, and bind to an unprivileged port, so I just removed this capability with `setcap cap_net_bind_service-ep /usr/bin/nodejs` and now GRPC_TRACE is working as expected.
